Implement FEFaceEvaluation::evaluate() for ECL/unstructured meshes

peterrum opened the pull request:
depends on #13056

peterrum commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -2928,9 +2933,42 @@ namespace internal
+
+              for (unsigned int i = 0; i < 3 * n_components * dofs_per_face;
+                   ++i)
+                temp[i][v] = temp[i][v]; // TODO: we need a second buffer?
```
Does `scratch_data` have enough space to use as buffer her?

kronbichler replied:
`3 * n_components * dofs_per_face` is exactly what we assume as memory before we let `scratch_data` begin. Since that additional array is long enough, we should have plenty of more space. But I guess this assignment is still to be updated as it is currently a self-assignment.

peterrum commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -2928,9 +2933,42 @@ namespace internal
+            MatrixFreeFunctions::DoFInfo::dof_access_cell &&
+          eval.get_is_interior_face() == false)
+        {
+          for (unsigned int v = 0; v < Number::size(); ++v)
+            {
```
I guess it would be better to loop over all `2*d` faces here, instead of looping over each lanes!?

kronbichler replied:
I guess it depends how often we have a particular face number and change in orientation. I would believe that the current choice might be more efficient for moderate amounts of irregularity, as the other alternative needs to mask out some of the lanes (I think of the results in the hanging nodes algorithms).

kronbichler replied:
Looking at this again, I believe I did not state my point correctly: It is likely that it would help to detect the case where all faces have the same number, and then go to the `else` branch below, and only run the per-lane code in the other case. Alternatively, we might want to detect how many of the faces actually pop up, and then only run the code for those cases. That would be strictly less work than the loop over all lanes, as we can't have more different faces than lanes. I guess this was what you had in mind?

kronbichler commented:
I can have a closer look once this is rebased.

peterrum commented:
I have rebased this PR.

kronbichler reviewed (approved):
Apart from the question of how we want to handle the lane divergence, I think this is good to go.

peterrum commented:
@kronbichler I have updated the PR, pushed some fixes and added tests. What about merging like this. I would do the vectorization check in a follow-up PR if that is fine?

peterrum commented:
/rebuild

kronbichler commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -2861,15 +2861,20 @@ namespace internal
         {
-          const std::size_t n_dofs     = shape_info.dofs_per_component_on_cell;
-          const std::size_t n_q_points = shape_info.n_q_points_faces[face_no];
+          Assert((fe_eval.get_dof_access_index() ==
+                    MatrixFreeFunctions::DoFInfo::dof_access_cell &&
+                  fe_eval.is_interior_face() == false) == false,
```
Can we negate this statement?
```suggestion
          Assert(fe_eval.get_dof_access_index() !=
                    MatrixFreeFunctions::DoFInfo::dof_access_cell ||
                  fe_eval.is_interior_face() == true,
```
